### ISSUE TITLE
Pithos randomly locks up during buffering.

### ISSUE THREAD

**Issue by JasonLG1979** [pithos/pithos#143]
PIthos randomly locks up while buffering to the point that it must be manually killed. I have experienced this in Fedora 21 and Ubuntu 14.10.

Normally GST_DEBUG nor Pithos are any help debugging this. Output just stops when Pithos locks up. But after putting a "try except:" in "start_song" I was finally able to get this.

```
ERROR - pithos:on_gst_error:678 - Gstreamer error: gst-resource-error-quark: Error while writing to download file. (10), gstdownloadbuffer.c(1244): gst_download_buffer_chain (): /GstPlayBin:player/GstURIDecodeBin:uridecodebin35/GstDownloadBuffer:downloadbuffer35:
Error writing file: Resource temporarily unavailable, 10
ERROR - pithos:on_gst_error:678 - Gstreamer error: gst-stream-error-quark: Internal data flow error. (1), gstbasesrc.c(2933): gst_base_src_loop (): /GstPlayBin:player/GstURIDecodeBin:uridecodebin35/GstSoupHTTPSrc:source:
streaming task paused, reason error (-5), 1
```
Looks very similar to #113, They may be the same bug.

**Comment by clrn0979** [comment]
I have a similar problem with pithos. When I start pithos it starts to buffer 1 to 2 percent and then goes back to zero..  It will eventually start after about 2 to 3 mins and plays the song all the way through. When it gets to the next song it will do the buffer thing again another 2 to 3 mins before song starts to play. Any help with this would be appreciated..

**Comment by glennimoss** [comment]
I think the issue of continued/looping buffering is #113. Let's make this issue specifically about Pithos completely locking up during buffering.

On the issue of locking, I'm pretty sure it's caused by some kind of deadlock in Gstreamer. I'm still trying to track it down...

**Comment by JasonLG1979** [comment]
It's got to be a gstreamer bug. It doesn't seem to effect the version that ships with Ubuntu 14.04(as far as I can tell anyway), but it does effect Ubuntu 14.10 and 15.04 as well as Fedora 20 and 21.

**Comment by JasonLG1979** [comment]
It's also noteworthy that the version in Ubuntu 14.04 does not buffer mp3 streams to disk(even with the flag set) but the others do.

**Comment by glennimoss** [comment]
Thanks for the idea. I'll look in to that.

**Comment by JasonLG1979** [comment]
Although even if the download buffer flag is not enabled in the newer gstreamer versions it still deadlocks from time to time. 

My kinda crappy solution is to stick with Kubuntu 14.04 until the issue is sorted in gstreamer. I'd rather be using Fedora 21 but a broken Pithos is a deal breaker.

**Comment by TingPing** [comment]
@JasonLG1979 The issue is reported upstream?

**Comment by glennimoss** [comment]
If I can ever get anything concrete on it I will report it upstream... assuming it's actually something wrong with gstreamer and not something we're doing.

**Comment by JasonLG1979** [comment]
@TingPing 

No. The flow of thing you would think would be. User reports bug to app devs > dev determines if it's a bug in the app or an upstream component > dev reports and handles bug with upstream on behalf of their app.

**Comment by JasonLG1979** [comment]
@glennimoss
I've tried as little as I know what to do. I've disabled the download buffer and I also hacked up a way to use track time to change tracks instead of gstreamer bus signals and still nothing. I've even played with qAndora, a Qt based Pandora client that uses either vlc or gstreamer 0.1 as the audio back end and uses old Pithos code to connect to Pandora. There is no lockup problem with either gstreamer0.1 or vlc.

**Comment by JasonLG1979** [comment]
As a side note it might be interesting to be able to switch back and forth between gsreamer and vlc back ends in Pithos. It might also help rule out the deadlock as a Pithos issue.

**Comment by TingPing** [comment]
>User reports bug to app devs > dev determines if it's a bug in the app or an upstream component > dev reports and handles bug with upstream on behalf of their app.

Well thats usually unlikely to happen, so if anybody expects progress upstream should be informed.

**Comment by JasonLG1979** [comment]
Well, I'm in the process of a rough port of Pithos to a VLC audio backend using https://github.com/oaubert/python-vlc. Everything seems to work pretty good. Aside from being a VERY quick and dirty port with no error handling on the audio stream the only real user facing niggle is that the VLC bindings don't keep very accurate time so the track time may slip every now and then.

**Comment by glennimoss** [comment]
I have determined that it's definitely a deadlock in the gstreamer library. I intend to report it and maybe submit a patch upstream, but I'm not quite sure when I can get around to that...

**Comment by JasonLG1979** [comment]
Right on.

**Comment by glennimoss** [comment]
I still have to figure out how to submit the upstream patch to gstreamer, but in the meantime I've built a package with the fix and put it in my PPA. (https://launchpad.net/~glennimoss/+archive/ubuntu/ppa) Right now it's only built for Utopic (14.10) but it's trivial to get it to build for other releases. Please let me know if you want it for some other release, or if you want to know the fix to apply it to something non-Ubuntu.

**Comment by gregsheremeta** [comment]
You rock.

I'm on Fedora. Can you post a .patch file here with the gstreamer changes, so I can test?

**Comment by gregsheremeta** [comment]
"""
please submit patches (in git format-patch) format to bugzilla. File a new bug and then attach the patch to the bug.
http://bugzilla.gnome.org/enter_bug.cgi?product=GStreamer
"""

**Comment by glennimoss** [comment]
The fix is made to gst-plugins-base version 1.4.3. Assuming it hasn't changed drastically should probably be easy to figure out how to apply this to other versions.

I'm not sure that the fix is the right way to go about this. It's possible I introduced some other threading issue by not locking before checking the `group->stream_changed_pending` flag. However, the case where that flag is true never happened over tens of hours of playing, whereas the deadlock would often happen on the second or third song in. When I submit the upstream bug, I hope to get some feedback from someone who knows gstreamer better than me.

Instead of removing the unused `group->stream_changed_pending_lock`, I did consider using it to lock all interactions with the `group->stream_changed_pending` flag, but I'm not sure that it would be THAT much better.

```patch
 gst/playback/gstplaybin2.c | 27 ++++++++++++++-------------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/gst/playback/gstplaybin2.c b/gst/playback/gstplaybin2.c
index 5583de1..0302c1b 100644
--- a/gst/playback/gstplaybin2.c
+++ b/gst/playback/gstplaybin2.c
@@ -360,7 +360,6 @@ struct _GstSourceGroup
 
   gulong block_id;
 
-  GMutex stream_changed_pending_lock;
   gboolean stream_changed_pending;
 
   /* to prevent that suburidecodebin seek flushes disrupt playback */
@@ -1282,7 +1281,6 @@ init_group (GstPlayBin * playbin, GstSourceGroup * group)
   g_mutex_init (&group->lock);
 
   group->stream_changed_pending = FALSE;
-  g_mutex_init (&group->stream_changed_pending_lock);
 
   /* init combiners. The combiner is found by finding the first prefix that
    * matches the media. */
@@ -1331,7 +1329,6 @@ free_group (GstPlayBin * playbin, GstSourceGroup * group)
   group->text_sink = NULL;
 
   group->stream_changed_pending = FALSE;
-  g_mutex_clear (&group->stream_changed_pending_lock);
 
   g_slist_free (group->suburi_flushes_to_drop);
   group->suburi_flushes_to_drop = NULL;
@@ -2781,17 +2778,21 @@ gst_play_bin_handle_message (GstBin * bin, GstMessage * msg)
      * groups (because the application set a new uri in about-to-finish)
      * if the playsink queue still has buffers to play */
 
-    GST_SOURCE_GROUP_LOCK (group);
     pending = group->stream_changed_pending;
-
     if (pending) {
-      GST_DEBUG_OBJECT (playbin, "Storing buffering message from pending group "
-          "%p %" GST_PTR_FORMAT, group, msg);
-      gst_message_replace (&group->pending_buffering_msg, msg);
-      gst_message_unref (msg);
-      msg = NULL;
+      GST_SOURCE_GROUP_LOCK (group);
+
+      // Double check in case it changed before we got the lock.
+      if (group->stream_changed_pending) {
+        GST_DEBUG_OBJECT (playbin, "Storing buffering message from pending group "
+            "%p %" GST_PTR_FORMAT, group, msg);
+        gst_message_replace (&group->pending_buffering_msg, msg);
+        gst_message_unref (msg);
+        msg = NULL;
+      }
+      GST_SOURCE_GROUP_UNLOCK (group);
     }
-    GST_SOURCE_GROUP_UNLOCK (group);
+
   } else if (GST_MESSAGE_TYPE (msg) == GST_MESSAGE_ERROR) {
     /* If we get an error of the subtitle uridecodebin transform
      * them into warnings and disable the subtitles */
```

**Comment by JasonLG1979** [comment]
Awesome!!! I'll checkout the fix on Ubuntu 14.10.  Regardless if Gstreamer accepts your patch or not I'm going to close this because it's definitely an upstream bug and @glennimoss is handling it.

**Comment by scottkosty** [comment]
Was anything done upstream for this? If so, any bug report that we can link to here?
